### PR TITLE
Add Schema Validation to Round Trip Abstract Test

### PR DESCRIPTION
**Pull request by sladyn98**
This PR deals with the validation of the JSON schema before application.
Closes https://github.com/jenkinsci/configuration-as-code-plugin/issues/1094

### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [x ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ x] Please describe what you did
- [x ] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

**Comment by sladyn98** [comment]
Should we wait for the test_framework to be merged before we deal with this. Would help to keep the code much cleaner rather than to just import stuff now ? @timja

**Comment by timja** [comment]
> Should we wait for the test_framework to be merged before we deal with this. Would help to keep the code much cleaner rather than to just import stuff now ? @timja

yes no point doing till that's merged

**Comment by sladyn98** [comment]
@timja Once the test framework pr is merged we could get this resolved.

**Comment by sladyn98** [comment]
@timja now that the framework is merged, perhaps we could move this forward.

File: plugin/src/test/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java [hunk 1]
```diff
@@ -104,6 +104,9 @@ public void roundTripTest() throws IOException {
             String jenkinsConf = getResourceContent(resourcePath);
             assertConfigViaWebUI(jenkinsConf);
 
+            //Validate Schema before application
+            assertTrue(validateSchema(convertYamlFileToJson(this, "validJenkinsConfigurator.yml")));
```
Comment (timja) [inline plugin/src/test/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: this should be the configuration being applied and not a static file in our repo.
also the api has now changed you should be asserting that there's no errors

File: test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java [hunk 2]
```diff
@@ -106,6 +107,9 @@ public void roundTripTest() throws IOException {
             String jenkinsConf = getResourceContent(resourcePath);
             assertConfigViaWebUI(jenkinsConf);
 
+            //Validate Schema before application
+            assertThat(validateSchema(convertToJson(resourceContent)), empty());
```
Comment (timja) [inline test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: Looks fine we should test this change on plugins that use this test and only do it after the nested schema change is merged

Comment (sladyn98) [reply test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: Sounds cool +1

Comment (timja) [reply test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: build is failing btw

File: test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java [hunk 3]
```diff
@@ -108,7 +112,7 @@ public void roundTripTest() throws IOException {
             assertConfigViaWebUI(jenkinsConf);
 
             //Validate Schema before application
-            assertThat(validateSchema(convertToJson(resourceContent)), empty());
+            assertThat(validateSchema(new JSONObject(new JSONTokener(convertToJson(resourceContent)))), empty());
```
Comment (timja) [inline test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: why is this needed?

Comment (sladyn98) [reply test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: Because of the type expected by ValidateSchema.

Comment (sladyn98) [reply test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: We can use the function directly but would need some modifications.

Comment (timja) [reply test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: probably worth looking at? this seems more complex than expected? maybe another helper method?

Comment (sladyn98) [reply test-harness/src/main/java/io/jenkins/plugins/casc/misc/RoundTripAbstractTest.java]: Implemented

**Comment by timja** [comment]
builds failing and can you please file a downstream PR with another plugin that's using this test?

**Comment by sladyn98** [comment]
> builds failing and can you please file a downstream PR with another plugin that's using this test?

Yeah sure, but how do I find out which plugin uses this test ?

**Comment by timja** [comment]
> > builds failing and can you please file a downstream PR with another plugin that's using this test?
> 
> Yeah sure, but how do I find out which plugin uses this test ?

search github in the jenkinsci organisation

**Comment by sladyn98** [comment]
@timja Do I file the same PR downstream them?

**Comment by timja** [comment]
> @timja Do I file the same PR downstream them?

first you need to get a green build, then a release candidate (incremental version) of this PR will automatically get published and then you can file a PR using your version.

**Comment by sladyn98** [comment]
@timja The tests are passing locally on my system but failing here.

**Comment by timja** [comment]
merge master into the branch @sladyn98, ci is likely merging master in.

**Comment by sladyn98** [comment]
> merge master into the branch @sladyn98, ci is likely merging master in.

Yeah looks like it :/

**Comment by sladyn98** [comment]
@timja Retriggered Build :+1:

File: integrations/src/test/resources/io/jenkins/plugins/casc/MailerTest.yml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 unclassified:
   mailer:
     replyToAddress: do-not-reply@acme.org
-    smtpHost: smtp.acme.org
-    smtpPort: 4441
+    smtpHost: "smtp.acme.org"
+    smtpPort: "4441"
```
Comment (timja) [inline integrations/src/test/resources/io/jenkins/plugins/casc/MailerTest.yml]: this is not ideal, I think I looked at it earlier and was going to be difficult to fix though, worried about breaking PCT again with this 😢 

Thoughts @oleg-nenashev / @casz ?

Comment (casz) [reply integrations/src/test/resources/io/jenkins/plugins/casc/MailerTest.yml]: Quoting should not be needed indeed, best figure out why it fails without quoting

Comment (sladyn98) [reply integrations/src/test/resources/io/jenkins/plugins/casc/MailerTest.yml]: I guess the reason I quoted was because the schema validation failed, it's time we developed our own validation engine using hamcrest validation

**Comment by sladyn98** [comment]
@jetersen anything we can do to fix this or is it good to go as is ?
@timja mentioned https://github.com/jenkinsci/configuration-as-code-plugin/pull/1121#discussion_r373770374
Is it blocking ?

**Comment by timja** [comment]
Yes blocking till answered on why and if its fixable

**Comment by sladyn98** [comment]
Closing this and can be opened if needed later
